SignalServiceKit/src/Messages/OWSReadReceiptManager.m

```diff
@@ -0,0 +1,191 @@
+
+    _messageSender = messageSender;
+
+    _toLinkedDevicesReadReceiptMap = [NSMutableDictionary new];
+
+    OWSSingletonAssert();
+
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(databaseViewRegistrationComplete)
+                                                 name:kNSNotificationName_DatabaseViewRegistrationComplete
+                                               object:nil];
+
+    return self;
+}
+
+- (void)dealloc
+{
+    [[NSNotificationCenter defaultCenter] removeObserver:self];
+}
+
+- (void)databaseViewRegistrationComplete
+{
+    [self scheduleProcessing];
+}
+
+// Schedules a processing pass, unless one is already scheduled.
+- (void)scheduleProcessing
+{
+    DispatchMainThreadSafe(^{
+        @synchronized(self)
+        {
+            if ([TSDatabaseView hasPendingViewRegistrations]) {
+                return;
+            }
+            if (self.isProcessing) {
+                return;
+            }
+
+            self.isProcessing = YES;
+
+            // Process read receipts every N seconds.
+            //
+            // We want a value high enough to allow us to effectively deduplicate,
+            // read receipts without being so high that we risk not sending read
+            // receipts due to app exit.
+            const CGFloat kProcessingFrequencySeconds = 3.f;
+            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(kProcessingFrequencySeconds * NSEC_PER_SEC)),
+                dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0),
+                ^{
+                    [self process];
+                });
+        }
+    });
+}
+
+- (void)process
+{
+    @synchronized(self)
+    {
+        self.isProcessing = NO;
+
+        NSArray<OWSReadReceipt *> *readReceiptsToSend = [[self.toLinkedDevicesReadReceiptMap allValues] copy];
+        [self.toLinkedDevicesReadReceiptMap removeAllObjects];
+        if (readReceiptsToSend.count > 0) {
+            OWSReadReceiptsMessage *message = [[OWSReadReceiptsMessage alloc] initWithReadReceipts:readReceiptsToSend];
+
+            dispatch_async(dispatch_get_main_queue(), ^{
+                [self.messageSender sendMessage:message
+                    success:^{
+                        DDLogInfo(@"%@ Successfully sent %zd read receipt to linked devices.",
+                            self.tag,
+                            readReceiptsToSend.count);
+                    }
+                    failure:^(NSError *error) {
+                        DDLogError(@"%@ Failed to send read receipt to linked devices with error: %@", self.tag, error);
+                    }];
+            });
+        }
+    }
+}
+
+- (void)messageWasReadLocally:(TSIncomingMessage *)message;
+{
+    @synchronized(self)
+    {
+        NSString *threadUniqueId = message.thread.uniqueId;
+        OWSAssert(threadUniqueId.length > 0);
+
+        // Only groupthread sets authorId, thus this crappy code.
```

Comment:
nit: update comments per previous discussion. Only legacy 1:1 incoming messages will be lacking an author id. Should we move the logic into the message class?

Reply:
Done.  I DRY'd this up and applied the new method in one new place.